taichi/analysis/same_statements.cpp

```diff
@@ -97,22 +120,57 @@ class IRNodeComparator : public IRVisitor {
       same = false;
       return;
     }
+    auto other = other_node->as<Stmt>();
 
     // If two identical statements can have different values, return false.
-    if (check_same_value_ && !stmt->is_container_statement() &&
-        !stmt->common_statement_eliminable()) {
-      same = false;
-      return;
+    // TODO: two identical GlobalPtrStmts cannot have different values,
+    //  but GlobalPtrStmt::common_statement_eliminable() is false.
+    if (check_same_value_ && stmt != other && !stmt->is_container_statement() &&
```

Comment:
nit: there are five conditions here... Could you either a) add some comments explaining the checks, or b) pull out these conditions and assign them to variables with more meaningful names?